### ISSUE TITLE
chpldoc sphinx errors should be reported using '.chpl' line numbers

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16903]
Currently, when a `chpldoc` command fails because sphinx is issuing an error about the comment string not being formatted as expected, it does so in terms of the generated `.rst` file.  This can be confusing because the typical `chpldoc` user may have no idea about chpldoc's use of sphinx or underlying `.rst` files.  It would be better if `chpldoc` reported errors in terms of the user's .chpl file instead.

For example, running chpldoc on this file, where I've commented out a routine that I no longer want in my code base:

```chapel
/*                                                                              
proc foo() {                                                                    
  writeln("This is a commented out piece of code, not a chpldoc comment");      
}                                                                               
*/
proc bar() {
  writeln("This is procedure bar()");
}

bar();
```

generates the message:

```bash
Running Sphinx v3.2.1
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 2 source files that are out of date
updating environment: [new config] 2 added, 0 changed, 0 removed
reading sources... [ 50%] index
reading sources... [100%] modules/badComment.doc


Warning, treated as error:
/var/folders/[big-tmpdir-name]/source/modules/badComment.doc.rst:25:Definition list ends without a blank line; unexpected unindent.
error: building html output from chpldoc sphinx project
```

As a user, it's not clear to me what the `.rst` file is (and it doesn't exist anymore for me to inspect it), or what I would need to do to my source file to fix these errors.  (In a short program like the one here, it's pretty obvious that this file isn't ready for `chpldoc`, but if this pattern shows up in a much larger code base, it can be challenging to get to the source of the problem.

When I think about what it would take to do this, I wonder whether:
* sphinx has a way of mapping lines back to the original source locations or definitions, similar to `# line` directives in cpp
* we'd have to learn to parse comments to ensure their legality within `chpldoc` itself if not

**Comment by lydia-duncan** [comment]
#19216 has a variation of this issue.  I'm closing that for now, but it'd be good to add a test of it when this one is resolved

**Comment by jabraham17** [comment]
I took at look at implementing this, and I don't think its feasible to implement for chpldoc.

For reference, other tools have this exact same issue, for example https://github.com/sphinx-gallery/sphinx-gallery/issues/725. Anytime `rst` is generated by a tool, the errors end up being in the form of the generated code. The only solution I can find (and the one taken in that issue) is to add comments to the generated file saying "DON'T EDIT THIS". This doesn't help much in our case, as users never even see the generated file.

One possibility is the include our generated rst files with a bunch of extra information allowing us to map back to the source. For example, we can hijack the error output of sphinx to capture errors, and rewrite them before showing the to the user
```python
        import io, re
        class StdErrorShim(io.TextIOWrapper):
            def __init__(self, stream):
                super().__init__(stream)
                self._stream = stream

            def write(self, data):
                if m := re.match(r'/.*source/modules/(.*?)\.rst:(\d+):\s*(.*)', data):
                    # replace the path with the source file
                    pass

                self._stream.write(data)
                self.flush()

        sys.stderr.flush()
        original_stderr = sys.stderr
        sys.stderr = StdErrorShim(original_stderr)
        from sphinx.cmd.build import main
        sys.exit(main())
        sys.stderr.flush()
        sys.stderr = original_stderr
```

But this requires us to put some kind of comment everywhere saying what lines map back to what chapel lines (like the C preprocessor `#line`). This will make the generated rst much harder to read for a human. And, it will require a more complex parsing of chpldoc comments, because we can't just insert a comment on every single line (for example, we can't put comments inside of code blocks or interrupt directives).

In short, I think its not feasible to implement this feature right now, however valuable and useful it might be. A possible alternative solution would be to add error messages saying "Error in generated RST: use `--save-sphinx gen` to see the generated RST`

**Comment by lydia-duncan** [comment]
Disappointing but that seems fair, thanks for investigating it, Jade!

**Comment by bradcray** [comment]
Ditto.  I'm closing this since it sounds like there's not much we can do, or plan to do, here, as the author (but Jade gets the credit if anyone's keeping score :) ).
